### PR TITLE
Omit invalid nodes instead of returning false

### LINKED ISSUE #4: Filter "empty" strings
Hi,

I'd like to filter out "empty" strings, that add noise to my component composition.

Imagine the following HTML source:

```html
<react-component react-name="Accordion">
    <react-component react-name="AccordionItem" title="Title 1">
        <p>Some text</p>
    </react-component>
    <react-component react-name="AccordionItem" title="Title 2">
        <p>Some other text</p>
    </react-component>
</react-component>
```
I correctly get instances of Accordion and AccordionItem.
But within my Accordion component `this.props.children` returns the following:
```js
["↵    ", Object, "↵    ", Object, "↵"]
````
I want to strip out these white-space strings. I tried to do that, by adjusting my `isValidNode` method:
```js
function(node) {
    if(typeof node === 'object') {
        return !(node.type === 'text' && node.data.trim().replace(/^\n+$/g, '') === '')
    }
    return true;
};
```
But this leads to:
```js
[false, Object, false, Object, false]
```
What I want is:
```js
[Object, Object]
```
Any hints for me?

### PR DESCRIPTION
**Pull request by Regaddi**
Fixes #4 implicitly.

### CODE REVIEW

**Comment by aknuds1** [comment]
Thanks, will consider this.

**Comment by aknuds1** [comment]
Would you mind adding a test corresponding to your #4 case? That would help in ensuring that it's solved and that it doesn't reoccur.

**Comment by Regaddi** [comment]
I think that test case sums it up. I took 2 child <p>-tags to simplify my `.length` check. With one valid child only React does not create an array.

**Comment by aindlq** [comment]
Will it preserve whitespaces in p tag?

**Comment by Regaddi** [comment]
@aindlq This totally depends on your processingInstructions. In my case I check for non-informational whitespace text, that only produces noise in my React structure.

File: test/html-to-react-tests.js [hunk 1]
```diff
@@ -340,7 +340,7 @@ describe('Html2React', function () {
                 assert.equal(reactHtml, htmlExpected);
             });
 
-            it('should return false in case of invalid node', function () {
+            it('should return undefined in case of invalid node', function () {
```
Comment (aknuds1) [inline test/html-to-react-tests.js]: What I don't understand is why we need to change the public API? I'm not sure why the public API is to return false in the first place, but still can't we fix this bug while retaining the API?

File: test/html-to-react-tests.js [hunk 2]
```diff
@@ -349,7 +349,39 @@ describe('Html2React', function () {
...
+            });
+
+            it('should not affect unhandled whitespace', function () {
+                var htmlInput = '<div> <p></p> <p></p> </div>';
+
+                var processingInstructions = [{
+                    shouldProcessNode: function (node) { return true; },
+                    processNode: processNodeDefinitions.processDefaultNode,
+                }, ];
+                var reactComponent = parser.parseWithInstructions(htmlInput,
```
Comment (aknuds1) [inline test/html-to-react-tests.js]: Doesn't this equal `parser.parse`?

Comment (Regaddi) [inline test/html-to-react-tests.js]: You're totally right :) I just deepdived in my specific problem :D

File: test/html-to-react-tests.js (hunk 1 above)
Comment (aknuds1) [inline test/html-to-react-tests.js]: Not sure I understand. Do we *have to* change the public API to return undefined as opposed to false here, in order to fix this bug?

**Comment by Regaddi** [comment]
Hey @aknuds1,
I reworked this a bit. `traverseDom` will now return false again.
Nevertheless if a node is invalid, it won't be merged into `props.children`.

I reset the tests I originally changed back to the behavior before.

**Comment by aindlq** [comment]
Hm, I'm not sure that I understand why this change is necessary. Why not just use something like:

```
    {
      shouldProcessNode: isValidNode,
      processNode: skipNode,
    }

function skipNode(
  node: Node, children: Array<React.ReactNode>
): React.ReactElement<any> {
  return null;
}
```

it should work because there is already check for null children

**Comment by Regaddi** [comment]
I'm not sure, what `isValidNode` should be there for, when I can't use it to sort out invalid nodes.
When using default `isValidNode` (which always returns `true`!?) I have to set up another custom processing instruction in my `processingInstructions` array.
This is in my opinion not really convenient.
Did I miss any other functionality of `isValidNode`?

**Comment by aindlq** [comment]
sorry, `isValidNode` was an arbitrary name. This is just additional check for processingInstruction where you can filter out things that you don't want to see in react, e.g. empty spaces.

**Comment by Regaddi** [comment]
Ok, this leads me back to this PR, where I don't want to mess up `isValidNode`s API by returning `null`.
I expect to get elements skipped, when `isValidNode` returns `false`.

So this PR basically cleans up the processing of `isValidNode` inside `traverseDom` where a `false` return leads to skipped React children.

One could discuss, if an invalid root node returns `false`, `undefined` or `null`. This is currently left as is, `false`.

**Comment by aindlq** [comment]
ok, thinking one more time I think it is a good default behaviour.

**Comment by aknuds1** [comment]
This looks OK to me at least, in that the previous behaviour would return `false` for invalid children, which doesn't make much sense.

**Pull request merged**